src/MongoDB.Driver.Core/Core/Clusters/MultiServerCluster.cs

```diff
@@ -358,7 +358,8 @@ private ClusterDescription ProcessReplicaSetChange(ClusterDescription clusterDes
             clusterDescription = EnsureServers(clusterDescription, args.NewServerDescription, newServers);
 
             if (args.NewServerDescription.CanonicalEndPoint != null &&
-                !EndPointHelper.Equals(args.NewServerDescription.CanonicalEndPoint, args.NewServerDescription.EndPoint))
+                !EndPointHelper.Equals(args.NewServerDescription.CanonicalEndPoint, args.NewServerDescription.EndPoint) &&
+                args.NewServerDescription.Type != ServerType.ReplicaSetPrimary)
```

Comment:
I didn't validate how this change matches with the spec yet, but at least the reason message should be updated

Reply:
Updated message.
Done.